addons/product/models/product.py

```diff
@@ -369,8 +369,28 @@ def unlink(self):
         self.clear_caches()
         return res
 
-    def _filter_to_unlink(self, check_access=True):
-        return self
+    def _filter_to_unlink(self):
+        fields_to_product = [
+            field
+            for model in self.env.values()
+            if not model._abstract
+            for field in model._fields.values()
+            if field.type == 'many2one'
+            and field.comodel_name == 'product.product'
+            and field.store
+            and field.ondelete == 'restrict'
+        ]
+        # TODO: a single raw sql request could work faster
+        to_unlink = self
+        for field in fields_to_product:
+            if not to_unlink:
+                break
+            domain = [(field.name, 'in', to_unlink.ids)]
+            records = self.env[field.model_name].read_group(domain, [field.name], [field.name])
+            linked_product_ids = [group[field.name][0] for group in records]
+            to_unlink = to_unlink - self.browse(linked_product_ids)
```

Comment:
Try this:
<pre>
records = self.env[field.model_name].read_group(domain, ['product_ids:array_agg(%s)' % field.name], [field.name])
linked_product_ids = records and records[0]['product_ids'] or []
if linked_product_ids:
    to_unlink = to_unlink - self.browse(linked_product_ids)
</pre>